nimare/meta/mkda.py

```diff
@@ -90,24 +93,144 @@ def _fit(self, dataset):
-
-        ma_values = ma_values * self.weight_vec
-        of_values = np.sum(ma_values, axis=0)
+        self.weight_vec_ = weight_vec  # C x 1 array
+        assert self.weight_vec_.shape[0] == ma_values.shape[0]
+        ma_values = ma_values * self.weight_vec_
+        stat_values = self._compute_summarystat(ma_values)
+
+        # Determine null distributions for summary stat (OF) to p conversion
+        if self.null == "analytic":
+            raise ValueError("'analytic' null distribution estimation is not yet supported.")
+        else:
+            self._compute_null_empirical(ma_values, n_iters=self.n_iters)
+        p_values, z_values = self._summarystat_to_p(stat_values, method=self.null)
 
         images = {
-            "of": of_values,
+            "stat": stat_values,
+            "p": p_values,
+            "z": z_values,
         }
         return images
 
+    def _compute_summarystat(self, data):
+        """Compute OF scores from data.
+
+        Parameters
+        ----------
+        data : array, pandas.DataFrame, or list of img_like
+            Data from which to estimate ALE scores.
+            The data can be:
+            (1) a 1d contrast-len or 2d contrast-by-voxel array of MA values,
+            (2) a DataFrame containing coordinates to produce MA values,
+            or (3) a list of imgs containing MA values.
+
+        Returns
+        -------
+        stat_values : 1d array
+            OF values. One value per voxel.
+        """
+        if isinstance(data, pd.DataFrame):
+            ma_values = self.kernel_transformer.transform(
+                data, masker=self.masker, return_type="array"
+            )
+        elif isinstance(data, list):
+            ma_values = self.masker.transform(data)
+        elif isinstance(data, np.ndarray):
+            ma_values = data.copy()
+        else:
+            raise ValueError('Unsupported data type "{}"'.format(type(data)))
+
+        # OF is just a sum of MA values.
+        stat_values = np.sum(ma_values, axis=0)
+        return stat_values
+
+    def _compute_null_empirical(self, ma_maps, n_iters=10000):
+        """Compute uncorrected null distribution using empirical method.
+
+        Parameters
+        ----------
+        ma_maps : (C x V) array
+            Contrast by voxel array of MA values.
+
+        Notes
+        -----
+        This method adds one entry to the null_distributions_ dict attribute:
+        "empirical_null".
+        """
+        n_studies, n_voxels = ma_maps.shape
+        weight_vec = np.squeeze(self.weight_vec_)  # cannot have singleton
+        assert weight_vec.ndim == 1
+        null_distribution = np.zeros(n_iters)
+        for i_iter in range(n_iters):
+            # One random MA value per study
+            null_ijk = np.random.choice(np.arange(n_voxels), n_studies)
+            iter_ma_values = ma_maps[np.arange(n_studies), null_ijk]
+            # Calculate summary statistic
+            iter_ma_values *= weight_vec
```

Comment:
Is this re-weighing of the `iter_ma_values` necessary? I see the ma_values that are passed into this function are already weighted, correct? https://github.com/neurostuff/NiMARE/pull/372/files#diff-09e94548290cad221e601257340cd9acfc7d1e851de54756374671589b3a6d74R98.

If that's the case I think this is (at least part of) the reason for the poor performance of MKDADensity with ALEKernel.

Reply:
I think you're right. The `ma_values` are already weighted. I'll fix that. Thanks!

Reply:
That fixed the failed tests on my end, so hopefully those tests will now pass on CI.